src/mariadb/videx/ha_videx.cc

```diff
@@ -0,0 +1,992 @@
+	DBUG_ENTER("ha_videx::index_next");
+	rc= HA_ERR_WRONG_COMMAND;
+	DBUG_RETURN(rc);
+}
+
+/**
+read backwards through the index. Not required for VIDEX.
+*/
+
+int ha_videx::index_prev(uchar *buf)
+{
+	int rc;
+	DBUG_ENTER("ha_videx::index_prev");
+	rc= HA_ERR_WRONG_COMMAND;
+	DBUG_RETURN(rc);
+}
+
+/**
+asks for the first key in the index. Not required for VIDEX.
+*/
+int ha_videx::index_first(uchar *buf)
+{
+	int rc;
+	DBUG_ENTER("ha_videx::index_first");
+	rc= HA_ERR_WRONG_COMMAND;
+	DBUG_RETURN(rc);
+}
+
+/**
+asks for the last key in the index. Not required for VIDEX.
+*/
+int ha_videx::index_last(uchar *buf)
+{
+	int rc;
+	DBUG_ENTER("ha_videx::index_last");
+	rc= HA_ERR_WRONG_COMMAND;
+	DBUG_RETURN(rc);
+}
+
+/**
+rnd_init() is called when the system wants the storage engine to do a table
+scan. Not required for VIDEX.
+*/
+int ha_videx::rnd_init(bool scan)
+{
+	DBUG_ENTER("ha_videx::rnd_init");
+	DBUG_RETURN(0);
+}
+
+int ha_videx::rnd_end()
+{
+	DBUG_ENTER("ha_videx::rnd_end");
+	DBUG_RETURN(0);
+}
+
+/**
+This is called for each row of the table scan. Not required for VIDEX.
+*/
+int ha_videx::rnd_next(uchar *buf)
+{
+	int rc;
+	DBUG_ENTER("ha_videx::rnd_next");
+	rc= HA_ERR_END_OF_FILE;
+	DBUG_RETURN(rc);
+}
+
+/**
+This is like rnd_next, but you are given a position to use
+to determine the row. Not required for VIDEX.
+*/
+int ha_videx::rnd_pos(uchar *buf, uchar *pos)
+{
+	int rc;
+	DBUG_ENTER("ha_videx::rnd_pos");
+	rc= HA_ERR_WRONG_COMMAND;
+	DBUG_RETURN(rc);
+}
+
+/**
+position() is called after each call to rnd_next() if the data needs
+to be ordered. Not required for VIDEX.
+*/
+void ha_videx::position(const uchar *record)
+{
+	DBUG_ENTER("ha_videx::position");
+	DBUG_VOID_RETURN;
+}
+
+/**
+Returns table statistics to the server; fills fields in the handler object.
+@return 0 on success or HA_ERR_*.
+*/
+
+int ha_videx::info(uint flag)
+{
+	return (info_low(flag, false));
+}
+
+/**
+Provides extra hints to the handler. Not required for VIDEX.
+@return 0 or error code.
+*/
+
+int ha_videx::extra(enum ha_extra_function operation)
+{
+	DBUG_ENTER("ha_videx::extra");
+	DBUG_RETURN(0);
+}
+
+/**
+MySQL calls this method at the end of each statement */
+
+int ha_videx::reset()
+{
+	DBUG_ENTER("ha_videx::reset");
+	DBUG_RETURN(0);
+}
+
+/**
+External lock per table; no-op for VIDEX.
+@return 0. */
+
+int ha_videx::external_lock(THD *thd, int lock_type)
+{
+	DBUG_ENTER("ha_videx::external_lock");
+	DBUG_RETURN(0);
+}
+
+/**
+Converts a MySQL table lock in 'lock' to the engine representation.
+Not required for VIDEX.
+@return pointer to the current element in 'to'.
+*/
+
+THR_LOCK_DATA **ha_videx::store_lock(THD *thd,
+                                       THR_LOCK_DATA **to,
+                                       enum thr_lock_type lock_type)
+{
+	if (lock_type != TL_IGNORE && lock.type == TL_UNLOCK)
+		lock.type=lock_type;
+	*to++= &lock;
+	return to;
+}
+
+/**
+Estimates the number of index records in a range via VIDEX HTTP.
```

Comment:
This is a very important function and represents a key modification introduced by VIDEX. We can describe it in more detail, like:

1. Working principle: VIDEX will forward requests via HTTP to the `VIDEX-Statistic-Server` (launched in a RESTful manner). If the request fails, a default value of 10 will be returned.

2. `Reference link`: For an implementation of VIDEX-Statistic-Server, see `https://github.com/bytedance/videx`. We plan to introduce it into the official MariaDB repository in a subsequent PR.